File: Algoritmos/python/puntofijo.py
"""
Método de Punto Fijo para encontrar raíces
Convierte f(x) = 0 en x = g(x) y aplica x_{n+1} = g(x_n)
"""


def punto_fijo(g, x0, tol=1e-6, max_iter=100):
    """
    Método de Punto Fijo

    Parámetros:
    -----------
    g : function
        Función de iteración g(x) tal que x = g(x)
    x0 : float
        Aproximación inicial
    tol : float
        Tolerancia para el criterio de parada
    max_iter : int
        Número máximo de iteraciones

    Retorna:
    --------
    raiz : float
        Aproximación de la raíz
    iteraciones : list
        Historial de iteraciones
    """
    iteraciones = []
    x = x0

    for i in range(max_iter):
        x_nuevo = g(x)

        iteraciones.append({
            'n': i,
            'x': x,
            'g(x)': x_nuevo,
            'error': abs(x_nuevo - x)
        })

        # Criterio de convergencia
        if abs(x_nuevo - x) < tol:
            print(f"Convergencia alcanzada en {i + 1} iteraciones")
            return x_nuevo, iteraciones

        x = x_nuevo

    print(f"No se alcanzó convergencia en {max_iter} iteraciones")
    return x, iteraciones


# Ejemplo de uso
if __name__ == "__main__":
    import math

    # Ejemplo: Encontrar raíz de x^2 - 2 = 0
    # Reescribimos como x = sqrt(2*x - x^2 + 2) o x = (x^2 + 2)/2x
    # O más simple: x = sqrt(2)  -> g(x) = (x + 2/x)/2
    g = lambda x: (x + 2 / x) / 2

    raiz, hist = punto_fijo(g, x0=1.0)

    if raiz:
        print(f"\nRaíz encontrada: {raiz:.10f}")
        print(f"Verificación: sqrt(2) = {math.sqrt(2):.10f}")

        print("\nHistorial de iteraciones:")
        for it in hist[-5:]:  # Últimas 5 iteraciones
            print(f"n={it['n']}: x={it['x']:.10f}, error={it['error']:.2e}")

File: Algoritmos/matlab/puntofijo.m
function [raiz, iteraciones] = punto_fijo(g, x0, tol, max_iter)
% PUNTO_FIJO Método de Punto Fijo para encontrar raíces
%
% Sintaxis: [raiz, iteraciones] = punto_fijo(g, x0, tol, max_iter)
%
% Parámetros:
%   g        - Handle de la función de iteración g(x)
%   x0       - Aproximación inicial
%   tol      - Tolerancia (opcional, default: 1e-6)
%   max_iter - Máximo de iteraciones (opcional, default: 100)
%
% Retorna:
%   raiz        - Aproximación de la raíz
%   iteraciones - Matriz con historial [n, x, g(x), error]

    % Valores por defecto
    if nargin < 3, tol = 1e-6; end
    if nargin < 4, max_iter = 100; end

    % Inicialización
    x = x0;
    iteraciones = zeros(max_iter, 4);

    for i = 1:max_iter
        x_nuevo = g(x);
        error = abs(x_nuevo - x);

        % Guardar iteración
        iteraciones(i, :) = [i, x, x_nuevo, error];

        % Criterio de convergencia
        if error < tol
            fprintf('Convergencia alcanzada en %d iteraciones\n', i);
            raiz = x_nuevo;
            iteraciones = iteraciones(1:i, :);
            return;
        end

        x = x_nuevo;
    end

    fprintf('No se alcanzó convergencia en %d iteraciones\n', max_iter);
    raiz = x;
    iteraciones = iteraciones(1:max_iter, :);
end

% Ejemplo de uso:
% g = @(x) (x + 2/x) / 2;
% [raiz, hist] = punto_fijo(g, 1.0);
% fprintf('Raíz encontrada: %.10f\n', raiz);